lib/orbit/transformable.js

```diff
@@ -1,121 +1,118 @@
+    data: transform,
+    process: function() {
+      return processTransform.call(_this, transform);
+    }
+  });
+
+  return action.complete;
 }
 
-function queueTransformation(transformation) {
+function queueTransformResult(transform, result) {
   var _this = this;
 
-  var processor = this._transformationQueue.push({
-    data: transformation,
+  var action = this._transformQueue.push({
+    data: transform,
     process: function() {
-      return transformation.process();
+      return settleTransform.call(_this, transform, result);
     }
   });
 
-  return processor;
+  return action.complete;
 }
 
 var Transformable = {
   extend: function(object, actions) {
     if (object._transformable === undefined) {
       object._transformable = true;
-      object._transformationQueue = new ActionQueue();
+      object._transformQueue = new ActionQueue();
+      object._transformLog = {};
 
       Evented.extend(object);
 
-      object.didTransform = function(operation, inverse) {
-        var normalized = normalize(operation);
-        var transformation = transformationFor.call(this, normalized);
-        if (transformation) {
-          // console.log('Transformable#didTransform - matching transformation found', this.id, normalized, inverse);
-          transformation.pushCompletedOperation(normalized, inverse);
-
-        } else {
-          // console.log('Transformable#didTransform - createTransformation', this.id, normalized, inverse);
-          transformation = new Transformation(this);
-          transformation.pushCompletedOperation(normalized, inverse);
-          queueTransformation.call(this, transformation);
-        }
+      object.didTransform = function() {
+        Orbit.deprecate('`didTransform` has been deprecated. Please call `transformed` instead and supply a TransformResult as an argument.', true);
       };
 
-      object.currentTransformation = function() {
-        if (this._transformationQueue.current) return this._transformationQueue.current.data;
-      };
+      object.transformed = function(result) {
+        assert('Call `transformed` with an instanceof a TransformResult.', result instanceof TransformResult);
 
-      object.transform = function(operation) {
-        var normalized = normalize(operation);
-        var transformation = transformationFor.call(this, normalized);
-        var action;
+        var transform = new Transform(result.operations);
 
-        if (transformation) {
-          // console.log('transform - matching transformation found', this.id, normalized);
-          action = transformation.pushOperation(normalized);
+        // console.log('transformed - queued', this.id, transform, result);
+        return queueTransformResult.call(this, transform, result);
+      };
 
-          if (isArray(action)) {
-            return action[action.length - 1].complete;
-          } else {
-            return action.complete;
-          }
+      object.transform = function(transform) {
```

Comment:
As you've said, the challenge here is that there is no connection between the source's application of a transform and firebase's acknowledgment of its application.

Given this, I think you have two options:

* Optimistic - assume success from firebase, and allow `transform` to succeed immediately. It will then trigger `didTransform` with the original transform as an argument as well as an empty result. Any subsequent transforms received from firebase should call `transformed` with a differential between the requested transforms and the actually performed ones.

* Pessimistic - figure out an async hook that can watch the firebase updates and match them up with applied transforms. Only when one matches up, do you resolve `_transform` with the result. Any other updates received will need to propagate with calls to `transformed`.

It might be helpful to have a simplified version of your source in the test suite so that we can model and test these behaviors directly. I think both the optimistic and pessimistic scenarios are possible given the current API, but both have their challenges. 

Also, I'm curious about the methodologies used in other firebase implementations, such as the adapter for ember-data.